plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestTestingPostgreSqlServer.java

```diff
@@ -61,42 +59,42 @@ public void tearDown()
+
+        postgreSqlServer.execute(sql);
+        assertEventually(() -> assertThat(postgreSqlServer.getRemoteDatabaseEvents()).contains(event));
     }
 
     @Test(timeOut = 60_000)
     public void testCapturingCancelledStatement()
-            throws Throwable
+            throws Exception
     {
         String sql = "SELECT pg_sleep(60)";
-        Set<RemoteDatabaseEvent> remoteDatabaseEvents = captureRemoteEventsDuring(() -> {
-            try (Connection connection = DriverManager.getConnection(postgreSqlServer.getJdbcUrl(), postgreSqlServer.getProperties());
-                    Statement statement = connection.createStatement()) {
-                Future<Boolean> executeFuture = threadPool.submit(() -> statement.execute(sql));
-                Thread.sleep(5000);
-                statement.cancel();
-                assertThatThrownBy(executeFuture::get)
-                        .hasRootCauseInstanceOf(SQLException.class)
-                        .hasRootCauseMessage("ERROR: canceling statement due to user request");
-            }
-            catch (SQLException e) {
-                throw new RuntimeException(e);
-            }
-        });
-        assertThat(remoteDatabaseEvents).contains(new RemoteDatabaseEvent(sql, RUNNING));
-        assertThat(remoteDatabaseEvents).contains(new RemoteDatabaseEvent(sql, CANCELLED));
-    }
 
-    private Set<RemoteDatabaseEvent> captureRemoteEventsDuring(CheckedRunnable runnable)
-            throws Throwable
-    {
-        List<RemoteDatabaseEvent> before = postgreSqlServer.getRemoteDatabaseEvents();
-        runnable.run();
-        List<RemoteDatabaseEvent> after = postgreSqlServer.getRemoteDatabaseEvents();
-        return Sets.difference(ImmutableSet.copyOf(after), ImmutableSet.copyOf(before));
+        // verify query was run before
+        RemoteDatabaseEvent running = new RemoteDatabaseEvent(sql, RUNNING);
+        RemoteDatabaseEvent cancelled = new RemoteDatabaseEvent(sql, CANCELLED);
+        assertThat(postgreSqlServer.getRemoteDatabaseEvents()).doesNotContain(running, cancelled);
+
+        try (Connection connection = DriverManager.getConnection(postgreSqlServer.getJdbcUrl(), postgreSqlServer.getProperties());
+                Statement statement = connection.createStatement()) {
+            Future<Boolean> executeFuture = threadPool.submit(() -> statement.execute(sql));
+
+            // wait for query to become running
+            assertEventually(() -> assertThat(postgreSqlServer.getRemoteDatabaseEvents()).contains(running));
+
+            // cancel the query
+            statement.cancel();
+            assertThatThrownBy(executeFuture::get)
+                    .hasRootCauseInstanceOf(SQLException.class)
+                    .hasRootCauseMessage("ERROR: canceling statement due to user request");
+        }
+
+        assertEventually(() -> assertThat(postgreSqlServer.getRemoteDatabaseEvents()).contains(cancelled));
```

Comment:
does it mean that `io.trino.plugin.jdbc.BaseJdbcConnectorTest#getRemoteDatabaseEvents` callers are flaky too?

Reply:
Not sure what you mean by flaky. Method does not fail, but as I said above events are asynchronous. So you might need to wait for your event.

That method is already used inside `assertEventually`.